onnxruntime/core/providers/webgpu/shader_variable.h

```diff
@@ -0,0 +1,338 @@
+  ShaderUsage(uint32_t usage) : usage{usage} {}
+
+  explicit operator bool() {
+    return usage != None;
+  }
+};
+
+// A helper class to make it easier to generate shader code related to indices calculation.
+class ShaderIndicesHelper {
+ public:
+  ShaderIndicesHelper(std::string_view name, ProgramVariableDataType type, ShaderUsage usage, const TensorShape& dims);
+
+  ShaderIndicesHelper(ShaderIndicesHelper&&) = default;
+  ShaderIndicesHelper& operator=(ShaderIndicesHelper&&) = default;
+
+  // get the number of components of the variable.
+  inline int NumComponents() const { return num_components_; }
+
+  // get the rank of the indices.
+  inline int Rank() const;
+
+  // create a WGSL expression ({varname}_indices_t) for getting indices from offset.
+  // \param offset: a WGSL expression (u32) representing the offset.
+  inline std::string OffsetToIndices(std::string_view offset_expr) const;
+
+  // create a WGSL expression (u32) for getting offset from indices.
+  // \param indices: a WGSL expression ({varname}_indices_t) representing the indices.
+  inline std::string IndicesToOffset(std::string_view indices_expr) const;
+
+  // create a WGSL expression (u32) for getting original offset from broadcasted indices.
+  // \param indices: a WGSL expression ({broadcasted_result_varname}_indices_t) representing the broadcasted indices.
+  // \param broadcasted_result: the broadcasted result variable.
+  inline std::string BroadcastedIndicesToOffset(std::string_view indices_expr, const ShaderIndicesHelper& broadcasted_result) const;
+
+  // create a WGSL expression ({varname}_indices_t) as an indices literal
+  // \param init: a list of indices values.
+  template <typename... TIndices>
+  inline std::string Indices(TIndices&&... indices_args) const;
+
+  // create a WGSL statement for setting value of the specified dimension of the indices.
+  // \param indices_var: name of the indices variable ({varname}_indices_t).
+  // \param idx: the index (i32|u32) of the dimension to set.
+  // \param value: the value (u32) to set.
+  template <typename TIdx, typename TVal>
+  inline std::string IndicesSet(std::string_view indices_var, const TIdx& idx_expr, const TVal& value) const;
+
+  // create a WGSL expression (u32) for getting value of the specified dimension of the indices.
+  // \param indices_var: name of the indices variable ({varname}_indices_t).
+  // \param idx: the index (i32|u32) of the dimension to get.
+  template <typename TIdx>
+  inline std::string IndicesGet(std::string_view indices_var, const TIdx& idx_expr) const;
+
+ protected:
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(ShaderIndicesHelper);
+
+  void Impl(std::ostream& ss) const;
+
+  std::string_view IndicesType() const;
+
+  std::string name_;
+  ProgramVariableDataType type_;  // for variable
+  int num_components_;            // for variable
+  int rank_;
+  TensorShape dims_;
+
+  mutable ShaderUsage usage_;
+  mutable std::set<const ShaderIndicesHelper*> broadcasted_to_;
```

Comment:
Who fill this set? Who own the pointers? Suggest adding a comment here saying the pointers should be non-null.

Reply:
The current code flow is like this:

Any kernel implementation need to implement `Program::GenerateShaderCode(ShaderHelper& shader) const;` for the specific derived Program class. The passed in instance of `ShaderHelper` can be used to create instance of `IndicesHelper` by using `ShaderHelper::AddInput()`,  `ShaderHelper::AddOutput()` or  `ShaderHelper::AddIndices()`.

Current code made an assumption: when any kernel implementation calls `ShaderIndicesHelper::BroadcastedIndicesToOffset`, the 2 indices helpers are generated from the shader helper. If this is the case, then the `broadcasted_to_` contains valid pointers.